### DISCUSSION TITLE
Compiling cuStateVec with CMake

### DISCUSSION THREAD

**Discussion by TysonRayJones** [NVIDIA/cuQuantum#85] [Q&A] [answered]
Hi there,

I am incorporating cuStateVec into a C++ project as a dynamic library using CMake, and although it works, I'm apprehensive I'm not using best practice. For context, I've installed `cuquantum-cuda-11` via `apt-get` on Ubuntu 20.04 as per [these instructions](https://developer.nvidia.com/cuquantum-downloads?target_os=Linux&target_arch=x86_64&Distribution=Ubuntu&target_version=20.04&target_type=deb_local), which has created (among many others) relevant files:
- `/usr/include/custatevec.h`, which `#include`'s existing files in `/usr/local/cuda/include/` (such as `library_types.h`)
- `/usr/lib/x86_64-linux-gnu/libcuquantum/11/libcustatevec.so` (and `.a`, etc)

I have an _existing_ CMake project which uses CUDA (without cuStateVec). Since it intends to support (very) older CMake versions, the root `CMakeLists.txt` does _not_ declare `CUDA` as a language, i.e. via `project(myProject LANGUAGES CUDA)`, but instead does an old fashioned:
```CMake
find_package(CUDA REQUIRED)
cuda_add_library(myProject mySourceFiles)
target_link_libraries(QuEST ${CUDA_LIBRARIES})
```
Already inadvisable of course; please forgive me, I'm stuck with it at the moment!

I now wish to adapt this to include cuStateVec. I _expected_ it to be as simple as replacing the final line above with:
```CMake
find_library(CUQUANTUM_LIBRARIES custatevec)
target_link_libraries(myProject ${CUDA_LIBRARIES} ${CUQUANTUM_LIBRARIES})
```
Alas, compiling yields the below error
```bash
/usr/include/custatevec.h:134:10: fatal error: library_types.h: No such file or directory
  134 | #include <library_types.h>
```
indicating that CMake has not included directory `/usr/local/cuda/include/`.
I can hackily remedy this by adding:
```CMake
target_include_directories(myProject PUBLIC "/usr/local/cuda/include")
```
and everything compiles fine, but I'm doubtful this is really addressing the issue.

So; is there an example of compiling cuStateVec with CMake, or additional `CMakeLists.txt` files I've overlooked?
It's possible the issue is really with my understanding of CMake, and irrelevant to cuStateVec's build, but I'm surprised I didn't have to point other CUDA libraries to the headers in `/usr/local/cuda/include`.

**Comment by mtjrider** [comment]
Hi @TysonRayJones thanks for posting this question.
We appreciate that projects have various legacy support requirements and appreciate the situation.

If any of my explanations below are not helpful, please let us know/clarify. Apologies for my long answer.

Without seeing exactly what is being compiled or how, it's difficult to comment on best practices. Can we migrate the best-practices question to a discussion, and can you provide more detail, there? One comment on this without knowing anything specific about your use-case, I'd recommend isolating and compiling artifacts that specifically use CUDA separately and linking them to your target executable/library at the end.

Regarding your other question about include directories: we intend cuQuantum be used like any other CUDA math library, and dependent projects of that library be compiled with Nvidia's toolchain (e.g.) `nvcc`. See our [example makefile](https://github.com/NVIDIA/cuQuantum/blob/c202ee7ad23ce9a5072a32b45d36e8310a20eeb5/samples/custatevec/Makefile#L22).

`nvcc` will automatically inject CUDA headers, flags, definitions, etc., where needed. You can see what's being injected by increasing verbosity or by performing a dry-run. [Here's some documentation](https://docs.nvidia.com/cuda/cuda-compiler-driver-nvcc/index.html#compilation-phases) on `nvcc` compilation phases.

If using `nvcc` to compile certain components is not an option, you can usually get away with adding the headers/flags manually, but I wouldn't generally recommend it. [Here's a chart](https://docs.nvidia.com/cuda/cuda-installation-guide-linux/index.html#host-compiler-support-policy) discussing host-compiler support.

**Comment by mtjrider** [comment]
@TysonRayJones let us know if this is still an issue.

**Comment by TysonRayJones** [comment]
Hi @mtjrider ,

Sure, feel free to migrate to a discussion if preferred!
I'm afraid though I haven't yet solved the issue, nor understand what would distinguish custatevec from other CUDA libraries to break this cmake build. I'll report my progress, though won't have a chance to try again for a short while. Happy to migrate this in the meantime.

**Reply by mtjrider** [reply]
Hi @TysonRayJones 

If it's helpful, I can post a concrete example of what I mean.

Let me know if you have outstanding questions.

**Comment by mtjrider** [comment]
@TysonRayJones 

Thanks. I've migrated the issue to a discussion, as nothing is "broken" per se. 
I assume you're compiling cuStateVec features into QuEST? Is there any specific build code / cmake you can share with me so that I can provide more specific/relevant information?

Ideally, we'd have a minimalistic reproducer to iterate over.

**Reply by TysonRayJones** [reply, marked as answer]
Apologies for the profane delay! Oliver Brown managed to [get this working](https://github.com/QuEST-Kit/QuEST/commit/522353076045334df58b7d2e40ed4224e33d222e) in QuEST v4 with a custom `FindCuQuantum.cmake` script, which is used in conjunction with a `CMakeLists.txt` containing:
```cmake
find_package(CUQUANTUM REQUIRED)
target_link_libraries(MyTarget PRIVATE CUQUANTUM::cuStateVec)
set(CMAKE_INSTALL_RPATH_USE_LINK_PATH ON)
  ```
I suspect this is as good as it gets until a more official `FindCuQuantum` is prepared!
